### DISCUSSION TITLE
Beginner Questions

### DISCUSSION THREAD

**Discussion by ianmrose-cm** [diesel-rs/diesel#3292] [Q&A] [answered]
Hello, I'm working on creating an initial setup for Diesel with Postgres, and I have some confusion on how the full setup is supposed to look based on the intro docs on the website. I am new to Diesel, Rust, and fairly new even to running a SQL database (I know the basics of how it works and such, but no significant experience), so that probably has something to do with it, but I think my use case is a fairly basic one and I've been having to go back and forth on documentation sources quite a bit to see if Diesel will work for me.

This is a fair bit of stuff so I can split into multiple discussions if desired. My questions are mostly around:
1. Managing rust definition alongside the SQL migration scripts
2. Handling of enums
3. Definitions using nested structs

Questions I have:

The [getting started guide](https://diesel.rs/guides/getting-started.html) starts by creating migration scripts to setup/teardown the database changes. This seems useful, but it gives off the impression that it is necessary for me to create a SQL version of the table definition in addition to modeling my tables as structs in my source code. Is that the case? 

I found [an example](https://github.com/diesel-rs/diesel/tree/master/examples/postgres/custom_types/src) under the postgres examples for this repo which seems to show that the schema can be generated purely from source. If so, should I even write a SQL version which needs to be kept manually in sync with what the source can output? This would seem to be fairly brittle. 

It makes sense to me that creating database migration scripts should be its' own separate thing which is not automatically run based on the source code, since that could break things, but in that case wouldn't the better workflow be to always generate schema from source and then have a tool to generate the diff between schemas as a SQL migration script? This would avoid the need to write SQL at all unless the table already exists and needs an initial migration. (in my case it is a new table & database so I don't need to) Generating the SQL is probably a fair bit of work so it makes sense it may not exist or may not even be worth it, but I think more guidance on how this process should work might help.

There don't seem to be any defaults for handling enums, but enums are fairly common when it comes to defining code structures. I realize the way enums are serialized for database storage can vary by database & by use case, but would it make sense to create some "sensible defaults" for this? I would think capability to store an enum as a string (VARCHAR or TEXT) would be a decent default approach since it makes the DB entries more readable than something like converting to an int. For now it looks like I can write the "toSql" and "fromSql" implementations for my enums, but it seems to all be boilerplate which can be eliminated. (I noticed there is some library to handle enums for diesel, but doesn't seem worth it to me to add another dependency just for that)

Nested structs are also hard to get a grip on from the intro materials. I actually don't see any examples of this within the [postgres examples folder](https://github.com/diesel-rs/diesel/tree/master/examples/postgres), though there were some interesting things in the more advanced example. Defining a table using nested structs seems like very common use case, so it would be great to have something which demonstrates how it will be converted into a table schema. Perhaps there is a clear way this should happen for people who are familiar with SQL database structures, but for me I'm not sure if a nested structure would result in:
1. A column which has multiple sub-columns/divisons (if this is a thing)
4. Each sub-struct is expanded into the same number of columns as its' fields, and perhaps the column name is prefixed to distinguish between clashing field names.
5. Something else

Example which would be similar to my use case (is somewhat random and incomplete, but has the basic functionality I need):
```
#[primary_key(internal_id)]
#[derive(Clone, Debug, Queryable)]
#[diesel(table_name = group)]
pub struct GroupTable {
    pub internal_id: i32,
    pub group: Group,
    pub perms: Perms
}

#[derive(Clone, Debug, Queryable)]
pub struct Group {
    pub id: String,
    pub members: Vec<String>,
    pub status: Status,
}

// need to write to/from sql
pub enum Status {
    Active,
    Inactive,
    Pending,
}

#[primary_key(internal_id)]
#[derive(Clone, Debug, Queryable)]
#[diesel(table_name = user)]
pub struct UserTable {
    pub internal_id: i32,
    pub id: String,
    pub name: String,
    pub membership: Membership,
}

#[derive(Clone, Debug, Queryable)]
pub struct Perms {
    pub read: Vec<String>,
    pub write: Vec<String>,
    pub owner: String,
}
```

I will be working on this in any case and will probably find the answers I need to move forward, but I think a lot of what I've described here could be called out or added within the main documentation to help with ramp up. Generating the migration scripts based on differences in the generated table schemas might be a nice tool, or maybe it is impractical (I don't know enough about how the schemas compare to SQL).

Thanks for your time.

**Comment by weiznich** [comment, marked as answer]
> It makes sense to me that creating database migration scripts should be its' own separate thing which is not automatically run based on the source code, since that could break things, but in that case wouldn't the better workflow be to always generate schema from source and then have a tool to generate the diff between schemas as a SQL migration script? This would avoid the need to write SQL at all unless the table already exists and needs an initial migration. (in my case it is a new table & database so I don't need to) Generating the SQL is probably a fair bit of work so it makes sense it may not exist or may not even be worth it, but I think more guidance on how this process should work might help.

It work the other way around. You write the migrations and diesel_cli generates the corresponding `schema.rs` representation based on a database where these migrations where applied. This has the major advantage that this approach also works with existing database and does not assume that you control the database at all.
That written: It might be possible to also generate migrations based on diff between the `schema.rs` file and your database, but that's something that's not implemented yet. I encourage people to experiment with this. If this turns out to be a good approach I'm open to integrate something like this in `diesel_cli` as well.

> There don't seem to be any defaults for handling enums, but enums are fairly common when it comes to defining code structures. I realize the way enums are serialized for database storage can vary by database & by use case, but would it make sense to create some "sensible defaults" for this? I would think capability to store an enum as a string (VARCHAR or TEXT) would be a decent default approach since it makes the DB entries more readable than something like converting to an int. For now it looks like I can write the "toSql" and "fromSql" implementations for my enums, but it seems to all be boilerplate which can be eliminated. (I noticed there is some library to handle enums for diesel, but doesn't seem worth it to me to add another dependency just for that)

The support for enums is by design as minimal as possible. Our general rule for including something in diesel is that it must be common practice and there must be a single way to solve the problem. That's not the case for enums, as enums can be represented in different ways for different database systems (custom enum types for postgresql, text columns, integer columns, …). Each use case has different requirements so we don't want to restrict them. By implementing `FromSql`/`ToSql` manually users are able to choose the correct approach for their implementation. In addition that allows third parties to provide opinionated implementations and removes maintenance burden from diesel itself. If you don't feel that adding a dependency for this is a good solution that's fine, as you always con provide such a generalized implementation by yourself.

> Nested structs are also hard to get a grip on from the intro materials. I actually don't see any examples of this within the postgres examples folder, though there were some interesting things in the more advanced example. Defining a table using nested structs seems like very common use case, so it would be great to have something which demonstrates how it will be converted into a table schema. Perhaps there is a clear way this should happen for people who are familiar with SQL database structures, but for me I'm not sure if a nested structure would result in:
>
> 1.   A column which has multiple sub-columns/divisons (if this is a thing)
> 2.   Each sub-struct is expanded into the same number of columns as its' fields, and perhaps the column name is prefixed to distinguish between clashing field names.
> 3.   Something else

It's both 1. and 2. You can construct a composite type by providing a corresponding `FromSql` impl that turns a database entry into a struct or your can provide a select clause that is structured in the same way as your struct. See [this documentation](https://docs.diesel.rs/1.4.x/diesel/pg/types/sql_types/struct.Record.html) for an example sql side type for the first variant. For the second variant you would need something like `.select((col1, col2, (col3, col4))` which can be mapped to some struct like this (assuming the field types match):

```rust
#[derive(Queryable)]
struct Foo {
    a: i32,
    b: i32,
    c: InnerFoo,
}

#[derive(Queryable)]
struct InnerFoo {
    something: i32,
    other_field: i32,
}
```

Generally speaking `#[derive(Queryable)]` maps query results literally by structure onto the corresponding result type. The first field of a tuple is mapped to the first field of the struct and so on. If you are using one of the 2.0.0 RC versions, you might be also interested in [`Selectable`](https://docs.diesel.rs/master/diesel/prelude/trait.Selectable.html) in this context.

**Reply by ianmrose-cm** [reply]
Nice, thanks for the comprehensive response so quickly!

I guess for me I was hoping the rust definitions of structs & marking them as tables would generate the schemas so that I don't need to worry about keeping my SQL definition of the table & my Rust structs in sync manually. Perhaps it isn't that difficult to keep them in sync, but I do get the point that the goal is to work with any database & not just one which is self-owned. (though I think it would still work for db's which aren't self-owned)

The point about enums makes sense, shouldn't be hard to write the to/from sql and I can probably get rid of the boilerplate pretty easily once I'm more familiar with Rust.

**Reply by weiznich** [reply]
> Perhaps it isn't that difficult to keep them in sync, but I do get the point that the goal is to work with any database & not just one which is self-owned. (though I think it would still work for db's which aren't self-owned)

Well at least diesel forces you normally to keep them in sync. Executing a migration will update your `schema.rs` file. This file is used to check your queries against, so if your database schema changes in incompatible ways there will be a compile time error pointing to the "broken" queries.
